File: dwarf-lib/runtime/py/result.py
"""Dwarf Result type for Python."""
from typing import Generic, TypeVar, Union

T = TypeVar("T")
E = TypeVar("E")
U = TypeVar("U")
F = TypeVar("F")


class _Ok(Generic[T, E]):
    def __init__(self, value: T) -> None:
        self._value = value

    def is_ok(self) -> bool:
        return True

    def is_err(self) -> bool:
        return False

    def unwrap(self) -> T:
        return self._value

    def unwrap_or(self, default: T) -> T:
        return self._value

    def map(self, fn):
        return ok(fn(self._value))

    def map_err(self, fn):
        return ok(self._value)

    def and_then(self, fn):
        return fn(self._value)

    def __repr__(self) -> str:
        return f"Ok({self._value!r})"


class _Err(Generic[T, E]):
    def __init__(self, error: E) -> None:
        self._error = error

    def is_ok(self) -> bool:
        return False

    def is_err(self) -> bool:
        return True

    def unwrap(self) -> T:
        raise ValueError(f"Called unwrap on Err: {self._error}")

    def unwrap_or(self, default: T) -> T:
        return default

    def map(self, fn):
        return err(self._error)

    def map_err(self, fn):
        return err(fn(self._error))

    def and_then(self, fn):
        return err(self._error)

    def __repr__(self) -> str:
        return f"Err({self._error!r})"


Result = Union[_Ok[T, E], _Err[T, E]]


def ok(value: T) -> _Ok[T, E]:
    return _Ok(value)


def err(error: E) -> _Err[T, E]:
    return _Err(error)


def is_ok(result: Result[T, E]) -> bool:
    return result.is_ok()


def is_err(result: Result[T, E]) -> bool:
    return result.is_err()

File: dwarf-lib/runtime/ts/result.ts
/**
 * Dwarf Result type — represents success (Ok) or failure (Err).
 *
 * Provides monadic operations (map, mapErr, andThen) as well as
 * standard unwrapping and type-narrowing helpers.
 *
 * @module
 */

export type Result<T, E> = Ok<T, E> | Err<T, E>;

export class Ok<T, E> {
  readonly tag = "Ok" as const;
  constructor(public readonly value: T) {}

  isOk(): this is Ok<T, E> { return true; }
  isErr(): this is Err<T, E> { return false; }

  unwrap(): T { return this.value; }
  unwrapOr(_defaultValue: T): T { return this.value; }

  map<U>(fn: (value: T) => U): Result<U, E> { return ok(fn(this.value)); }
  mapErr<F>(_fn: (err: E) => F): Result<T, F> { return ok(this.value) as unknown as Result<T, F>; }

  andThen<U>(fn: (value: T) => Result<U, E>): Result<U, E> { return fn(this.value); }
}

export class Err<T, E> {
  readonly tag = "Err" as const;
  constructor(public readonly error: E) {}

  isOk(): this is Ok<T, E> { return false; }
  isErr(): this is Err<T, E> { return true; }

  unwrap(): never { throw new Error(`Called unwrap on Err: ${this.error}`); }
  unwrapOr<U>(defaultValue: U): U { return defaultValue; }

  map<U>(_fn: (value: T) => U): Result<U, E> { return err(this.error) as unknown as Result<U, E>; }
  mapErr<F>(fn: (err: E) => F): Result<T, F> { return err(fn(this.error)); }

  andThen<U>(_fn: (value: T) => Result<U, E>): Result<U, E> { return err(this.error) as unknown as Result<U, E>; }
}

/**
 * Wraps a value in `Ok`.
 */
export function ok<T, E = never>(value: T): Ok<T, E> {
  return new Ok(value);
}

/**
 * Wraps an error in `Err`.
 */
export function err<T = never, E = unknown>(error: E): Err<T, E> {
  return new Err(error);
}

/**
 * Type guard — returns `true` if the result is `Ok`.
 */
export function isOk<T, E>(result: Result<T, E>): result is Ok<T, E> {
  return result.isOk();
}

/**
 * Type guard — returns `true` if the result is `Err`.
 */
export function isErr<T, E>(result: Result<T, E>): result is Err<T, E> {
  return result.isErr();
}
